pkg/raft/ready.go

```diff
@@ -0,0 +1,288 @@
+// This code has been modified from its original form by The Cockroach Authors.
+// All modifications are Copyright 2024 The Cockroach Authors.
+//
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package raft
+
+import (
+	"fmt"
+	"iter"
+	"strings"
+
+	pb "github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+)
+
+type SnapshotStatus int
+
+const (
+	SnapshotFinish  SnapshotStatus = 1
+	SnapshotFailure SnapshotStatus = 2
+)
+
+var emptyState = pb.HardState{}
+
+// SoftState provides state that is useful for logging and debugging.
+// The state is volatile and does not need to be persisted to the WAL.
+type SoftState struct {
+	RaftState pb.StateType
+}
+
+func (a *SoftState) equal(b *SoftState) bool {
+	return a.RaftState == b.RaftState
+}
+
+// StorageAppend describes a storage append request. It contains the updates
+// that must be written to storage atomically, such as new HardState, snapshot,
+// and/or log entries. When the updates are durable on the log storage, the
+// attached messages can be sent.
+type StorageAppend struct {
```

Comment:
Elaborate more on requrements:

- All `StorageAppend`s must be written in the order they are issued.
- `StorageAppendAck`s can be stepped in any order (which allows for some concurrency in their delivery).